GoogleUtilities/SceneDelegateSwizzler/GULSceneDelegateSwizzler.m

```diff
@@ -0,0 +1,466 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "TargetConditionals.h"
+
+#import <GoogleUtilities/GULAppEnvironmentUtil.h>
+#import <GoogleUtilities/GULLogger.h>
+#import <GoogleUtilities/GULMutableDictionary.h>
+#import <GoogleUtilities/GULSceneDelegateSwizzler.h>
+#import "../Common/GULLoggerCodes.h"
+#import "Internal/GULSceneDelegateSwizzler_Private.h"
+
+#import <objc/runtime.h>
+
+#if UISCENE_SUPPORTED
+
+API_AVAILABLE(ios(13.0), tvos(13.0))
+typedef void (*GULOpenURLContextsIMP)(id, SEL, UIScene *, NSSet<UIOpenURLContext *> *);
+
+API_AVAILABLE(ios(13.0), tvos(13.0))
+typedef void (^GULSceneDelegateInterceptorCallback)(id<UISceneDelegate>);
+
+// The strings below are the keys for associated objects.
+static char const *const kGULRealIMPBySelectorKey = "GUL_realIMPBySelector";
+static char const *const kGULRealClassKey = "GUL_realClass";
+
+static GULLoggerService kGULLoggerSwizzler = @"[GoogleUtilities/SceneDelegateSwizzler]";
+
+// Since Firebase SDKs also use this for app delegate proxying, in order to not be a breaking change
+// we disable App Delegate proxying when either of these two flags are set to NO.
+
+/** Plist key that allows Firebase developers to disable App and Scene Delegate Proxying. */
+static NSString *const kGULFirebaseSceneDelegateProxyEnabledPlistKey =
+    @"FirebaseAppDelegateProxyEnabled";
+
+/** Plist key that allows developers not using Firebase to disable App and Scene Delegate Proxying.
+ */
+static NSString *const kGULGoogleUtilitiesSceneDelegateProxyEnabledPlistKey =
+    @"GoogleUtilitiesAppDelegateProxyEnabled";
+
+/** The prefix of the Scene Delegate. */
+static NSString *const kGULSceneDelegatePrefix = @"GUL_";
+
+/**
+ * This class is necessary to store the delegates in an NSArray without retaining them.
+ * [NSValue valueWithNonRetainedObject] also provides this functionality, but does not provide a
+ * zeroing pointer. This will cause EXC_BAD_ACCESS when trying to access the object after it is
+ * dealloced. Instead, this container stores a weak, zeroing reference to the object, which
+ * automatically is set to nil by the runtime when the object is dealloced.
+ */
+@interface GULSceneZeroingWeakContainer : NSObject
+
+/** Stores a weak object. */
+@property(nonatomic, weak) id object;
+
+@end
+
+@implementation GULSceneZeroingWeakContainer
+@end
+
+@implementation GULSceneDelegateSwizzler
+
+#pragma mark - Public methods
+
++ (BOOL)isSceneDelegateProxyEnabled {
```

Comment:
I wonder if we can use `return [GULAppDelegateSwizzler isAppDelegateProxyEnabled]` as an implementation of the method to avoid code duplication?

Reply:
I think we should keep AppDelegateSwizzler and SceneDelegateSwizzler separated. Reusing that method can be confusing.

Reply:
As for me, as far as the implementation is identical, we should reuse it in that or another way. A more or less graceful solution could be introducing a separate class or function that encapsulates the implementation and is used by both App Delegate and Scene Delegate swizzlers. Then we will have cleaner code and less tests duplication. 

The downside of this approach is that we would have to change well tested and critical for Analytics App Delegate Swizzler which may introduce another issue. Taking this into account, using `[GULAppDelegateSwizzler isAppDelegateProxyEnabled]` as the implementation of `isSceneDelegateProxyEnabled` looks like a reasonable tradeoff, so we have less code duplication but don't touch  `GULAppDelegateSwizzler` just before the release.

I am kind of ok with the current code, though it feels like we will definitely forget to update one of the methods when the flag potentially changes  in the future.

WDTY?

Reply:
Sounds good, I don't have a strong preference on this.